src/Time/Timestamp.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE Rank2Types                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
+module Time.Timestamp
+       ( Timestamp (..)
+       , timeDiff
+       , timeAdd
+       , timeMul
+       , timeDiv
+       ) where
+
+import GHC.Real (denominator, numerator, (%))
+
+import Time.Rational (KnownRat, RatioNat)
+import Time.Units (Time (..))
+
+-- | Similar to 'Time' but has no units and can be negative.
+newtype Timestamp = Timestamp Rational
+    deriving (Show, Read, Num, Eq, Ord, Enum, Fractional, Real, RealFrac)
+
+
+-- | Returns the result of comparison of two 'Timestamp's and
+-- the 'Time' of that difference of given time unit.
+timeDiff :: KnownRat unit => Timestamp -> Timestamp -> (Ordering, Time unit)
+timeDiff (Timestamp a) (Timestamp b) =
+    let order         = compare a b
+        dif           = abs (a - b)
+        t :: RatioNat = fromIntegral (numerator dif) % fromIntegral (denominator dif)
+    in (order, Time t)
+
+-- | Returns the result of addition of two 'Time' elements.
+timeAdd :: KnownRat unit => Time unit -> Time unit -> Time unit
+timeAdd = (+)
```

Comment:
How did you manage to use `(+)` here? Time *most definitely* cannot have a `Num` instance.

Reply:
https://github.com/serokell/o-clock/issues/14

Reply:
Personally, I'm okay with `Num` instance for `Time`. You can write `(3 :: Second) + 5` without problems. Also you can write `(5 :: Second) - 3`. If you out of range while subtracting, when `Natural` will throw exception. Not the perfect behavior, but still might be convenient, if you use it carefully.

`(3 :: Second) * (5 :: Second)` probably doesn't make sense from physics point of view. But let's user decide what they want: more convenient interface with `(*)` or more type-safe with `timeAdd`. Also, having `Num` instances allows us to have other numeric instances with `floor` functions which are useful in real life.

Reply:
`(3 :: Second) * (5 :: Second)` is ridiculous and no sane library for units would allow this.

Reply:
if you insist on having a `Num` instance for numeric literals, make multiplication and division throw an error.

Reply:
`time-units` allows this. And we never had problems with it (because we never used `*` for time units). It's just the `Num` hierarchy which is not perfect. Unfortunately, we can't have everything from `Num` except `(*)`. But the rest of `Num` looks really useful and convenient. So ¯\_(ツ)_/¯.

Having `error` inside `(*)` and `(/)` operators is okay to me.